### ISSUE TITLE
🌟💡 YOLOv5 Study: batch size

### ISSUE THREAD

**Issue by glenn-jocher** [ultralytics/yolov5#2377]
## Study 🤔

I did a quick study to examine the effect of varying batch size on YOLOv5 trainings. The study trained YOLOv5s on COCO for 300 epochs with `--batch-size` at 8 different values: `[16, 20, 32, 40, 64, 80, 96, 128]`.

We've tried to make the train code batch-size agnostic, so that users get similar results at any batch size. This means users on a 11 GB 2080 Ti should be able to produce the same results as users on a 24 GB 3090 or a 40 GB A100, with smaller GPUs simply using smaller batch sizes.

We do this by scaling loss with batch size, and also by scaling weight decay with batch size. At batch sizes smaller than 64 we accumulate loss before optimizing, and at batch sizes above 64 we optimize after every batch.

## Results 😃

Initial results vary significantly with batch size, but final results are nearly identical (good!).
<img width="1301" alt="Screen Shot 2021-03-05 at 1 22 03 PM" src="https://user-images.githubusercontent.com/26833433/110174990-dd487d00-7db5-11eb-9118-48777ecd57c0.png">

Closeup of mAP@0.5:0.95:
<img width="1109" alt="Screen Shot 2021-03-05 at 1 27 33 PM" src="https://user-images.githubusercontent.com/26833433/110175415-9018db00-7db6-11eb-888f-6400376923e1.png">

One oddity that stood out is val objectness loss, which did vary with batch-size. I'm not sure why, as val-box and val-cls did not vary much, and neither did the 3 train losses. I don't know what this means or if there's any room for concern (or improvement).
<img width="1109" alt="Screen Shot 2021-03-05 at 1 27 21 PM" src="https://user-images.githubusercontent.com/26833433/110175428-9444f880-7db6-11eb-8607-1de5a2f81f4e.png">

**Comment by abhiagwl4262** [comment]
@glenn-jocher May be when we train for large number of epochs then we don't see significant improvement. I did experiment for batch size of 32 and 48 and I got better result when I trained with larger Batch Size. I trained for 50 epochs. And it happened on multiple datasets.

**Comment by glenn-jocher** [comment]
@abhiagwl4262 we always recommend you train on the largest batch-size possible, not so much for better performance, as the above results don't indicate higher performance with higher batch size, but certainly for faster training and better resource utilization.

Multi-GPU may add another angle to the above story though, as larger batch sizes there may help contribute to better results, at least in early training, since the batchnorm stats are split there among your CUDA devices.

**Comment by abhiagwl4262** [comment]
@glenn-jocher Is High Batch good, even for very small dataset e.g 200 images per class ?

**Comment by glenn-jocher** [comment]
@abhiagwl4262 maybe, as long as you maintain a similar number of iterations. For very small datasets this may require significantly increasing training epochs, i.e. to several thousand, or until you observe overfitting.

**Comment by cszer** [comment]
Hey, good thing to study. But i need to notice that results with sync BN are not reproducible for me. I have trained yolo m model on 8 tesla a100 gpus with batch size 256 because ddp only supports gloo backend and 0 GPU was loaded 50% more than others. (cuda 11). It will be good to compare syncbn with BN training.

**Comment by glenn-jocher** [comment]
@cszer thanks for the comments! Yes a --sync study would be interesting as well. What you are your observations with and without --sync?

Excess CUDA device 0 memory usage was previously related to too-large batch sizes on device 0 when testing, but this bug was fixed on February 6th as part of PR https://github.com/ultralytics/yolov5/pull/2148. If your results are from before that then you may want to update your code and see if the problem has been fixed.

**Comment by cszer** [comment]
1-2 05:095 map lower on coco

**Comment by glenn-jocher** [comment]
@cszer oh wow, that's a significant difference. Do you mean that you see a drop of -1 or -2 mAP on COCO when not using `--sync-bn` on a 8x A100 YOLOv5m training at --batch 256? That's much larger than I would have expected. Did you train for 300 epochs?

**Comment by abhiagwl4262** [comment]
@glenn-jocher  One very strange observation. I am able to run 48 batch size on single GPU and not able to run batch size 64 even on 2 GPUs. Is there some bug in multi-GPU implementation ?

**Comment by glenn-jocher** [comment]
@abhiagwl4262 if you believe you have a reproducible problem, please raise a new issue using the 🐛 **Bug Report template**, providing screenshots and a [minimum reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) to help us better understand and diagnose your problem. Thank you!
